### ISSUE TITLE
Feature Request: Support translating patch rules to template in ClusterClass controller

### ISSUE THREAD

**Issue by Pangzhihong** [kubernetes-sigs/cluster-api#10556]
### What would you like to be added (User Story)?

Support translating patch rules to template in ClusterClass controller

### Detailed Description

The current clusterclass function in the cluster-api can apply a cluster's patch rule to a template, which was an internal method at the time. We have a scenario where we need to use this patch function, and we want to provide an ability to escape the content of the patch rule to the corresponding template.

### Label(s) to be applied

/kind feature
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by fabriziopandini** [comment]
/priority awaiting-more-evidence

Could you kindly provide more detail about what do you mean with "The current clusterclass function in the cluster-api" and what you want to achieve.

As a general note, please be aware that generally we don't want to expose CAPI controllers code as a library / public functions; there are only two exception:
- the main reconcile function is alway spublic, so folks can eventually bin pack controllers
- we expose carefully few functions for the sake of making it easier to implement E2E tests.

**Comment by Pangzhihong** [comment]
> /priority awaiting-more-evidence
> 
> Could you kindly provide more detail about what do you mean with "The current clusterclass function in the cluster-api" and what you want to achieve.
> 
> As a general note, please be aware that generally we don't want to expose CAPI controllers code as a library / public functions; there are only two exception:
> 
> * the main reconcile function is alway spublic, so folks can eventually bin pack controllers
> * we expose carefully few functions for the sake of making it easier to implement E2E tests.

We want to use this method(applyPatchesToRequest）to implement the patch capability of the template (instead of relying on CAPI controllers)，currently, this method is not public:
func applyPatchesToRequest(ctx context.Context, req *runtimehooksv1.GeneratePatchesRequest, resp *runtimehooksv1.GeneratePatchesResponse) error {}

**Comment by sbueringer** [comment]
Sorry but we're not going to make arbitrary internal functions public. Feel free to copy the code if it's useful to you. It's just not sustainable to us to maintain something like this.

**Comment by Pangzhihong** [comment]
> Sorry but we're not going to make arbitrary internal functions public. Feel free to copy the code if it's useful to you. It's just not sustainable for us to maintain something like this.

Can you provide relevant controllers to parse the clusterclass results into various templates and provide them to the Portal or Web to demonstrate the correctness of the patch rule？

**Comment by sbueringer** [comment]
No sorry. We can' add additional controllers to CAPI for this use case

**Comment by Pangzhihong** [comment]
> No sorry. We can't add additional controllers to CAPI for this use case

Thanks for the reply, but I'm wondering why，does it affect the definition of the CAPI architecture? If we want to implement such requirements, can we only copy the code?

**Comment by sbueringer** [comment]
We are usually not adding controllers to CAPI to provide a backend for UIs that folks are building on top of CAPI.

The goal of our controllers is to reconcile objects within Kubernetes clusters.

**Comment by sbueringer** [comment]
@fabriziopandini Maybe you have a better way to express it, but this seems entirely out of scope for the project

**Comment by fabriziopandini** [comment]
> if we want to implement such requirements, can we only copy the code?

Yes, the project uses Apache 2 licence, so you are free to copy the code.

> @fabriziopandini Maybe you have a better way to express it, but this seems entirely out of scope for the project

The reason is explained in https://cluster-api.sigs.k8s.io/user/manifesto#the-complexity-budget
TL;DR As a project we must focus on what is core to the project/important for the majority of the users. Transforming CAPI controllers into a library for others to re-use bit and pieces is not core to the project, and it is the reason why we moved all the controllers to internal packages back in the 1.0 times (with the only allowed exceptions I have already documented above)

Also, PTAL https://github.com/kubernetes-sigs/cluster-api/issues/10138 who document previous attempt to solve the same problem and why we are now dropping this effort (because there was no interest to implement it up to a good state)

Happy to discuss this at the office hours if it can help, but I agree with @sbueringer to not proceed with this feature request

**Comment by Pangzhihong** [comment]
> > if we want to implement such requirements, can we only copy the code?
> 
> Yes, the project uses Apache 2 licence, so you are free to copy the code.
> 
> > @fabriziopandini Maybe you have a better way to express it, but this seems entirely out of scope for the project
> 
> The reason is explained in https://cluster-api.sigs.k8s.io/user/manifesto#the-complexity-budget TL;DR As a project we must focus on what is core to the project/important for the majority of the users, and do it well. Transforming CAPI controllers into a library for others to re-use bit and pieces is not core to the project, and it is the reason why we moved all the controllers to internal packages back in the 1.0 times (with the only allowed exceptions I have already documented above)
> 
> Also, PTAL #10138 who document previous attempt to solve the same problem and why we are now dropping this effort (because there was no interest to implement it up to a good state)
> 
> Happy to discuss this at the office hours if it can help, but I agree with @sbueringer to not proceed with this feature request

Thanks, what do you guys suggest, folk to a personal repository and then copy it and use it or is there a better suggestion?

**Comment by fabriziopandini** [comment]
I think there is no perfect answers, it boils up to use the process you are comfortable with.

Since this is not going to be implemented in Cluster API
/close
